ddtrace/internal/_encoding.pyx

```diff
@@ -46,120 +47,327 @@ cdef inline int PyBytesLike_Check(object o):
+
+
+cdef inline int pack_text(msgpack_packer *pk, object text):
+    cdef Py_ssize_t L
+    cdef int ret
+
+    if text is None:
+        return msgpack_pack_nil(pk)
+
+    if PyBytesLike_Check(text):
+        L = len(text)
+        if L > ITEM_LIMIT:
+            PyErr_Format(ValueError, b"%.200s object is too large", Py_TYPE(text).tp_name)
+        ret = msgpack_pack_raw(pk, L)
+        if ret == 0:
+            ret = msgpack_pack_raw_body(pk, <char *> text, L)
+        return ret
+
+    if PyUnicode_Check(text):
+        IF PY_MAJOR_VERSION >= 3:
+            ret = msgpack_pack_unicode(pk, text, ITEM_LIMIT)
+            if ret == -2:
+                raise ValueError("unicode string is too large")
+        ELSE:
+            text = PyUnicode_AsEncodedString(text, "utf-8", NULL)
+            L = len(text)
+            if L > ITEM_LIMIT:
+                raise ValueError("unicode string is too large")
+            ret = msgpack_pack_raw(pk, L)
+            if ret == 0:
+                ret = msgpack_pack_raw_body(pk, <char *> text, L)
+        return ret
+
+    raise TypeError("Unhandled text type: %r" % type(text))
+
 
-    usage::
+cdef inline int array_prefix_size(int l):
+    if l < 16:
+        return 1
+    elif l < (2<<16):
+        return 3
+    return 5
 
-        packer = Packer()
-        astream.write(packer.pack_trace(trace))
-        astream.write(packer.pack_traces(traces))
 
-    Packer's constructor has some keyword arguments:
+cdef class StringTable(object):
+    cdef dict _index
 
-    :param callable default:
-        Convert user type to builtin type that Packer supports.
-        See also simplejson's document.
-    """
+    def __init__(self):
+        self._index = {"": 0}
+        self.insert("")
+
+    cpdef insert(self, object string):
+        pass
+
+    cpdef index(self, object string):
+        cdef int index
+
+        if not string:
+            return 0
+        
+        try:
+            return self._index[string]
+        except KeyError:
+            index = len(self)
+            self._index[string] = index
+            self.insert(string)
+            return index
+
+    def __len__(self):
+        return len(self._index)
+
+    def __contains__(self, object string):
+        return string in self._index
+
+
+cdef class MsgpackStringTable(StringTable):
     cdef msgpack_packer pk
-    cdef object _default
-    cdef const char *encoding
-    cdef const char *unicode_errors
+    cdef int max_size
+    cdef int _sp
+    cdef object _lock
 
-    def __cinit__(self):
-        cdef int buf_size = 1024*1024
-        self.pk.buf = <char*> PyMem_Malloc(buf_size)
+    def __init__(self, max_size):
+        self.pk.buf = <char*> PyMem_Malloc(max_size)
         if self.pk.buf == NULL:
             raise MemoryError("Unable to allocate internal buffer.")
-        self.pk.buf_size = buf_size
-        self.pk.length = 0
+        self.max_size = self.pk.buf_size = max_size
+        self.pk.length = 6
```

Comment:
Why do we initialize the packer length to have an offset of 6?

Reply:
we need to leave room at the front of the buffer to prepend the array header (1 byte to say it is an array and then 5 to say how big it is)